### PR TITLE
Filecoin Integration

### PR DESCRIPTION
**Pull request by asutula**
The main goal here is to get all the Rigs imagery stored on web3/nft.storage in a way that is compatible with the future integrated operation of IPFS and Filecoin that PL is currently working on (it's not done yet). This means storing each rig's imagery as a separate item (cid). Additionally we want to add some representation of the fact that the imagery is stored in Filecoin to the Rig metadata. 

Currently, all 3000 rig's imagery have been stored on nft.storage and there have been four storage deals completed for each set of imagery. I've tracked all those deals in `artifacts/local.db`. The metadata queries have been updated to represent the deals in the metadata attributes. Here's an example of the resulting metadata:

```
{
  "name": "Rig #100",
  "external_url": "https://garage.tableland.xyz/rigs/100",
  "image": "ipfs://bafybeih66fnjzmtmcgij2zwhqui4f7udqw2vtuk3adyyhx4hw3yf7qcu7i/image_full.png",
  "image_alpha": "ipfs://bafybeih66fnjzmtmcgij2zwhqui4f7udqw2vtuk3adyyhx4hw3yf7qcu7i/image_full_alpha.png",
  "image_medium": "ipfs://bafybeih66fnjzmtmcgij2zwhqui4f7udqw2vtuk3adyyhx4hw3yf7qcu7i/image_medium.png",
  "image_medium_alpha": "ipfs://bafybeih66fnjzmtmcgij2zwhqui4f7udqw2vtuk3adyyhx4hw3yf7qcu7i/image_medium_alpha.png",
  "thumb": "ipfs://bafybeih66fnjzmtmcgij2zwhqui4f7udqw2vtuk3adyyhx4hw3yf7qcu7i/image_thumb.png",
  "thumb_alpha": "ipfs://bafybeih66fnjzmtmcgij2zwhqui4f7udqw2vtuk3adyyhx4hw3yf7qcu7i/image_thumb_alpha.png",
  "animation_url": "https://rigs.tableland.xyz/100.html",
  "attributes": [
    {
      "display_type": "number",
      "trait_type": "% Original",
      "value": "47.5"
    },
    {
      "display_type": "string",
      "trait_type": "Background",
      "value": "Desat Canyonland 2"
    },
    {
      "display_type": "string",
      "trait_type": "Chassis",
      "value": "Midnight DDL"
    },
    {
      "display_type": "string",
      "trait_type": "Cockpit",
      "value": "Midnight Genesis Block"
    },
    {
      "display_type": "string",
      "trait_type": "Filecoin Deal 1",
      "value": "https://filfox.info/deal/36739326"
    },
    {
      "display_type": "string",
      "trait_type": "Filecoin Deal 2",
      "value": "https://filfox.info/deal/37000535"
    },
    {
      "display_type": "string",
      "trait_type": "Filecoin Deal 3",
      "value": "https://filfox.info/deal/38987515"
    },
    {
      "display_type": "string",
      "trait_type": "Filecoin Deal 4",
      "value": "https://filfox.info/deal/39012246"
    },
    {
      "display_type": "string",
      "trait_type": "Fleet",
      "value": "Hoppers"
    },
    {
      "display_type": "string",
      "trait_type": "Garage Status",
      "value": "parked"
    },
    {
      "display_type": "string",
      "trait_type": "Mod",
      "value": "Dawn Equi-join"
    },
    {
      "display_type": "string",
      "trait_type": "Propulsion",
      "value": "Sunset Bind Variable"
    },
    {
      "display_type": "string",
      "trait_type": "VIN",
      "value": "2dcdaf4cdd4b18459ab6323dcfc7d22e2a62af02b1dd4e30777a759be2d0a0d5"
    }
  ]
}
```

There are plenty of supporting changes to make that all possible and some small improvements I found and made along the way.

### CODE REVIEW

File: pkg/dirpublisher/dirpublisher.go [hunk 1]
```diff
@@ -219,3 +221,127 @@ L1:
...
+func (dp *DirPublisher) RigsIndexToWeb3Storage(ctx context.Context, rigs []local.Rig) (cid.Cid, error) {
+	n, err := qp.BuildMap(basicnode.Prototype.Any, int64(len(rigs)), func(ma datamodel.MapAssembler) {
+		for _, rig := range rigs {
+			c, err := cid.Decode(*rig.RendersCid)
+			if err != nil {
+				log.Default().Printf("error decoding rig cid string: %v", err)
+				continue
+			}
+			qp.MapEntry(ma, fmt.Sprintf("%d", rig.ID), qp.Link(cidlink.Link{Cid: c}))
+		}
```
Comment (asutula) [inline pkg/dirpublisher/dirpublisher.go:290]: Here I'm building up a map keyed by a simple id created from each `Rig`s `id`, and values that are cid links to some data previously uploaded to web3.storage.

File: pkg/dirpublisher/dirpublisher.go [hunk 2]
```diff
@@ -219,3 +221,127 @@ L1:
...
+	})
+	if err != nil {
+		return cid.Cid{}, fmt.Errorf("building map: %v", err)
+	}
+
+	reader, writer := io.Pipe()
```
Comment (asutula) [inline pkg/dirpublisher/dirpublisher.go:296]: Using a pipe here so I can simultaneously encode the map as `dagcbor` and upload it to web3.storage.

File: pkg/dirpublisher/dirpublisher.go [hunk 3]
```diff
@@ -219,3 +221,127 @@ L1:
...
+			ID:   2,
+			Desc: "upload ipld",
+			ExecFn: func(ctx context.Context) (interface{}, error) {
+				c, err := dp.web3Storage.PutCar(ctx, reader)
+				if err != nil {
+					return nil, fmt.Errorf("uploading ipld: %v", err)
+				}
+				return c, nil
+			},
+		},
```
Comment (asutula) [inline pkg/dirpublisher/dirpublisher.go:319]: These two jobs get run on separate goroutines by the code below. Job 1 encodes the map as `dagcbor` to the pipe writer. Job two simultaneously uploads the data to web3.storage using `PutCar` and the pipe reader.

Comment (asutula) [reply pkg/dirpublisher/dirpublisher.go:319]: I get an error from Job 2 `unexpected EOF`. I'm guessing that is coming from the web3.storage backend?

File: pkg/dirpublisher/dirpublisher.go [hunk 4]
```diff
@@ -219,3 +221,128 @@ L1:
...
+		case <-pool.Done:
+			break L
+		}
+	}
+	log.Default().Printf("%s uploaded to web3.storage", res.String())
+	return res, nil
+}
+
+// RigsIndexToWeb3Storage creates a dagcbor index from the provided Rigs and adds it to web3.storage.
+func (dp *DirPublisher) RigsIndexToWeb3Storage(ctx context.Context, rigs []local.Rig) (cid.Cid, error) {
```
Comment (asutula) [inline pkg/dirpublisher/dirpublisher.go:282]: @brunocalza and @jsign, I'm having some trouble in this code that writes some dagcbor and uploads it to web3.storage. Wondering if you can see anything wrong. You can follow my comments in the below function, it's pretty simple. Thanks.

File: pkg/publisher/publisher.go [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+		if r.ID == 2 {
+			res = r.Value.(cid.Cid)
+		}
+	}
+	log.Default().Printf("%s uploaded to car storage service", res.String())
+	return res, nil
+}
+
+// RigsIndexToCarStorage creates a dagcbor index from the provided Rigs and adds it to car storage service.
+func (p *Publisher) RigsIndexToCarStorage(ctx context.Context, rigs []local.Rig) (cid.Cid, error) {
```
Comment (asutula) [inline pkg/publisher/publisher.go:257]: Goal of this func is to create a car file that maps each rig id (`int`) to a cid link for imagery that was already added to web3.storage/nft.storage in a previous step.

File: pkg/publisher/publisher.go [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+	n, err := qp.BuildMap(basicnode.Prototype.Any, int64(len(rigs)), func(ma datamodel.MapAssembler) {
+		for _, rig := range rigs {
+			c, err := cid.Decode(*rig.RendersCid)
+			if err != nil {
+				log.Default().Printf("error decoding rig cid string: %v", err)
+				continue
+			}
+			qp.MapEntry(ma, fmt.Sprintf("%d", rig.ID), qp.Link(cidlink.Link{Cid: c}))
+		}
```
Comment (asutula) [inline pkg/publisher/publisher.go:266]: Loop through the rigs and build up the map.

File: pkg/publisher/publisher.go [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+			Desc: "upload ipld",
+			ExecFn: func(ctx context.Context) (interface{}, error) {
+				c, err := p.carStorage.PutCar(ctx, reader)
+				if err != nil {
+					return nil, fmt.Errorf("uploading ipld: %v", err)
+				}
+				return c, nil
+			},
+		},
+	}
```
Comment (asutula) [inline pkg/publisher/publisher.go:296]: This code spins off two concurrent jobs. One for encoding map as dagcbor and the other to upload the resulting data to web3.storage/nft.storage. `p.carStorage` is an interface type that abstracts both web3.storage and nft.storage implementations so I can easily switch between the two. Job 2 gives me an `Unexpected EOF` error when using web3.storage and an `Unexpected end of data` error when using nft.storage. Obviously I'm doing somthing wrong, but can't see what.

File: cmd/rigs/cmd/publish.go [hunk 8]
```diff
@@ -23,17 +26,18 @@ import (
...
+	store     storage.Store
+	tblClient *client.Client
+	chain     chains.Chain
 )
 
 func init() {
 	rootCmd.AddCommand(publishCmd)
 
 	publishCmd.PersistentFlags().String("nft-storage-key", "", "api key for nft.storage")
+	publishCmd.PersistentFlags().String("web3-storage-key", "", "api key for web3.storage")
```
Comment (asutula) [inline cmd/rigs/cmd/publish.go:40]: Added the option to use web3.storage in addition to nft.storage for storing imagery. There is a storage interface and implementations for both services.

File: cmd/rigs/cmd/publish.go [hunk 9]
```diff
@@ -80,8 +88,10 @@ func init() {
 	publishCmd.PersistentFlags().String("eth-api-url", "", "ethereum api url")
 	publishCmd.PersistentFlags().String("infura-key", "", "api key for Infura")
 	publishCmd.PersistentFlags().String("alchemy-key", "", "api key for Alchemy")
+	publishCmd.PersistentFlags().String("ankr-key", "", "api key for Ankr")
+	publishCmd.PersistentFlags().Bool("glif", false, "use Glif provider")
 
-	publishCmd.MarkFlagsMutuallyExclusive("eth-api-url", "infura-key", "alchemy-key")
+	publishCmd.MarkFlagsMutuallyExclusive("eth-api-url", "infura-key", "alchemy-key", "ankr-key", "glif")
```
Comment (asutula) [inline cmd/rigs/cmd/publish.go:94]: Options for Glif and Ankr so you can publish data to Tableland on FEVM.

File: cmd/rigs/cmd/publish_clear-data.go [hunk 10]
```diff
@@ -16,7 +16,9 @@ func init() {
 
 	clearDataCmd.Flags().Bool("parts", false, "clear data for the parts table")
 	clearDataCmd.Flags().Bool("layers", false, "clear data for the layers table")
+	clearDataCmd.Flags().Bool("rigs", false, "clear data for the rigs table")
```
Comment (asutula) [inline cmd/rigs/cmd/publish_clear-data.go:19]: Throughout the PR there are many mechanical changes like this just to accommodate the new rigs and deals tables. Don't worry too much about these types of changes.

File: cmd/rigs/cmd/publish_data.go [hunk 11]
```diff
@@ -56,35 +61,37 @@ var dataCmd = &cobra.Command{
...
-				fmt.Printf("processed job %d. %s\n", r.ID, r.Desc)
-			case <-pool.Done:
-				fmt.Println("done")
-				break Loop
+		for r := range pool.Results() {
+			if r.Err != nil {
+				fmt.Printf("error processing job %d: %v\n", r.ID, r.Err)
+				continue
 			}
+			fmt.Printf("processed job %d. %s\n", r.ID, r.Desc)
```
Comment (asutula) [inline cmd/rigs/cmd/publish_data.go:94]: Cleaned up the API for my worker pool utility so this loop is no longer necessary. So you'll see diffs like this all over the PR, don't worry about them.

File: cmd/rigs/cmd/publish_data.go [hunk 12]
```diff
@@ -181,28 +222,68 @@ func attrsJobs(ctx context.Context, s local.Store, jobID *int) ([]wpool.Job, err
...
 				LayersCid:            layersCid,
+				IndexCid:             indexCid,
 				ImageFullName:        "image_full.png",
 				ImageFullAlphaName:   "image_full_alpha.png",
 				ImageMediumName:      "image_medium.png",
 				ImageMediumAlphaName: "image_medium_alpha.png",
 				ImageThumbName:       "image_thumb.png",
 				ImageThumbAlphaName:  "image_thumb_alpha.png",
 				AnimationBaseURL:     "https://rigs.tableland.xyz/",
+				FilecoinBaseURL:      "https://filfox.info/deal/",
```
Comment (asutula) [inline cmd/rigs/cmd/publish_data.go:286]: This is the url that we end up using in the metadata query. Are we cool with pointing to Filfox or should we use Filescan or other?

File: cmd/rigs/cmd/publish_layers.go [hunk 13]
```diff
@@ -36,16 +36,17 @@ var layersCmd = &cobra.Command{
...
+			checkErr(localStore.TrackCid(ctx, "layers", c.String()))
 		}
 		if chunksDir == "" && c.Defined() {
-			chunksDir, err = dirPublisher.CidToCarChunks(ctx, c)
+			chunksDir, err = pub.CidToCarChunks(ctx, c)
 			checkErr(err)
 			fmt.Printf("Car chunks written to folder %s\n", chunksDir)
 		}
-		checkErr(dirPublisher.CarChunksToNftStorage(
+		checkErr(pub.CarChunksToCarStorage(
```
Comment (asutula) [inline cmd/rigs/cmd/publish_layers.go:49]: These changes and similar changes elsewhere are just to accommodate the refactor that abstracts interaction with web3.storage and nft.storage with an interface.

File: cmd/rigs/cmd/publish_renders_index.go [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/spf13/viper"
+)
+
+func init() {
+	rendersIndexCmd.Flags().String("out", "./index.car", "filename to write the index CAR to")
+
+	publishCmd.AddCommand(rendersIndexCmd)
+}
+
+var rendersIndexCmd = &cobra.Command{
```
Comment (asutula) [inline cmd/rigs/cmd/publish_renders_index.go:16]: New command that creates a CAR index of all the rig ids and their corresponding imagery cids. It's basically a map of rig id to cid. The root cid of the map can then be pinned or unpinned on any pinning service and all the cids in the map will also be pinned/unpinned. This is much easier than managing 3000 individual cids.

File: cmd/rigs/cmd/publish_status.go [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+	"fmt"
+
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	publishCmd.AddCommand(statusCmd)
+}
+
+var statusCmd = &cobra.Command{
```
Comment (asutula) [inline cmd/rigs/cmd/publish_status.go:14]: Just a little command to check the storage status of a cid in web3/nft.storage manually. Made it to debug my work and figured I'd leave it.

File: cmd/rigs/cmd/publish_update_deals.go [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+func init() {
+	updateDealsCmd.Flags().Int("concurrency", 10, "number of concurrent requests to car storage service")
+	updateDealsCmd.Flags().Duration("rate-limit", time.Millisecond*350, "rate limit for requests to car storage service")
+	updateDealsCmd.Flags().StringSlice("ids", []string{}, "limit to rigs with these ids")
+
+	publishCmd.AddCommand(updateDealsCmd)
+}
+
+var updateDealsCmd = &cobra.Command{
```
Comment (asutula) [inline cmd/rigs/cmd/publish_update_deals.go:19]: New command that pulls the latest storage status from web3/nft.storage and stores the latest state in `local.db`. You can optionally provide a list of rig ids if you want to limit the rigs to check.

File: ethereum/deployments.ts [hunk 17]
```diff
@@ -41,7 +43,9 @@ export const deployments: RigsDeployments = {
     allowlistTable: "rigs_allowlist_42161_14",
     partsTable: "parts_42161_7",
     layersTable: "layers_42161_8",
+    rigsTable: "",
     attributesTable: "rig_attributes_42161_15",
+    dealsTable: "",
```
Comment (asutula) [inline ethereum/deployments.ts:48]: New entries for the new rigs and deals tables. We'll need to fill these in for mainnet once we create those tables.

File: ethereum/helpers/uris.ts [hunk 18]
```diff
@@ -22,15 +24,15 @@ export async function getURITemplate(
         await normalize(
           `select 
             json_object(
-              'name',case when exists(select * from ${pilotsTable} where rig_id = ${attributesTable}.rig_id and end_time is null) then 'Rig #'||rig_id||' ✈️' else 'Rig #'||rig_id end,
```
Comment (asutula) [inline ethereum/helpers/uris.ts:25]: In here, I updates the metadata queries to use the new lookups table as well as add deal information to the attributes. I made some other changes to just improve the readability and simplicity of the queries.

File: pkg/publisher/publisher.go [hunk 19]
```diff
@@ -0,0 +1,390 @@
+package publisher
```
Comment (asutula) [inline pkg/publisher/publisher.go:1]: This package used to be call `dirpublisher`, so it contains lots of pre-existing code. Not much to worry about in here.

File: pkg/carstorage/web3storage/client.go [hunk 20]
```diff
@@ -0,0 +1,65 @@
+package web3storage
```
[the remark this answers is not in the document]
Comment (asutula) [reply pkg/carstorage/web3storage/client.go:1]: Much simpler because they have a proper Go client.

File: pkg/publisher/carindex.go [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/fluent/qp"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/multiformats/go-multicodec"
+	"github.com/multiformats/go-multihash"
+)
+
+// RendersIndexToCar creates a CAR index from the saved Rigs and stores it as a CAR file.
+func (p *Publisher) RendersIndexToCar(ctx context.Context, filename string) (cid.Cid, error) {
```
Comment (asutula) [inline pkg/publisher/carindex.go:22]: Code for creating the CAR index file.

File: pkg/storage/local/store.go [hunk 22]
```diff
@@ -31,12 +45,14 @@ type Layer struct {
...
 	ID                int     `json:"id"`
 	Original          bool    `json:"original"`
+	RendersCid        *string `json:"renders_cid" db:"renders_cid"`
 	PercentOriginal   float64 `json:"percent_original" db:"percent_original"`
 	PercentOriginal50 float64 `json:"percent_original_50" db:"percent_original_50"`
 	PercentOriginal75 float64 `json:"percent_original_75" db:"percent_original_75"`
 	PercentOriginal90 float64 `json:"percent_original_90" db:"percent_original_90"`
 	VIN               string  `json:"vin"`
 	Parts             []Part  `json:"parts"`
+	Deals             []Deal  `json:"deals"`
```
Comment (asutula) [inline pkg/storage/local/store.go:55]: More updates to model deals in our local data model that is written to our local SQLite db.

File: pkg/storage/tableland/common/sqlfactory.go [hunk 23]
```diff
@@ -55,6 +55,23 @@ func (s *SQLFactory) SQLForInsertingLayers(table string, layers []local.Layer) (
...
+		)
+	}
+	ds := s.d.Insert(table).Cols("id", "renders_cid").Vals(vals...)
+	sql, _, err := ds.ToSQL()
+	if err != nil {
+		return "", fmt.Errorf("creating sql to insert rigs: %v", err)
+	}
+	return clean(sql), nil
+}
+
```
Comment (asutula) [inline pkg/storage/tableland/common/sqlfactory.go:74]: Transform data stored in our local SQLite db into SQL insert statements to populate our new rigs table on Tableland.

File: pkg/storage/tableland/common/sqlfactory.go [hunk 24]
```diff
@@ -110,31 +127,68 @@ func (s *SQLFactory) SQLForInsertingRigAttributes(rigAttrTable string, rigs []lo
...
+	ds := s.d.Insert(table).
+		Cols("rig_id", "deal_id", "storage_provider", "data_model_selector", "deal_number").
+		Vals(vals...)
+	sql, _, err := ds.ToSQL()
+	if err != nil {
+		return "", fmt.Errorf("creating sql to insert deals: %v", err)
+	}
+	return clean(sql), nil
+}
+
```
Comment (asutula) [inline pkg/storage/tableland/common/sqlfactory.go:150]: Transform data stored in our local SQLite db into SQL insert statements to populate our new deals table on Tableland.

File: pkg/storage/tableland/common/sqlfactory.go [hunk 25]
```diff
@@ -110,31 +127,68 @@ func (s *SQLFactory) SQLForInsertingRigAttributes(rigAttrTable string, rigs []lo
...
+			"animation_base_url", lookups.AnimationBaseURL,
+		},
+		{
+			"filecoin_base_url", lookups.FilecoinBaseURL,
+		},
+	}
+	ds := s.d.Insert(lookupsTable).Cols(
+		"label",
+		"value",
+	).Vals(vals...)
```
Comment (asutula) [inline pkg/storage/tableland/common/sqlfactory.go:191]: Updates for the new lookups table schema.

File: pkg/storage/tableland/impl/tableland/store.go [hunk 26]
```diff
@@ -67,7 +67,7 @@ func (s *Store) CreateTable(ctx context.Context, definition tableland.TableDefin
 
 // InsertParts implements InsertParts.
 func (s *Store) InsertParts(ctx context.Context, parts []local.Part) error {
-	tableName, err := s.localStore.TableName(ctx, "parts", s.chainID)
```
Comment (asutula) [inline pkg/storage/tableland/impl/tableland/store.go:70]: Transform data stored in our local SQLite db into SQL insert statements to populate our new rigs table on Tableland.

File: pkg/storage/tableland/store.go [hunk 27]
```diff
@@ -37,6 +37,14 @@ var (
...
 		)`,
 	}
+	// RigsDefinition defines the rigs table.
+	RigsDefinition = TableDefinition{
+		Prefix: "rigs",
+		Schema: `(
+			id integer primary key,
+			renders_cid text not null
+		)`,
+	}
```
Comment (asutula) [inline pkg/storage/tableland/store.go:47]: Definition for the new rigs Tableland table.

File: pkg/storage/tableland/store.go [hunk 28]
```diff
@@ -48,19 +56,24 @@ var (
...
+		Prefix: "deals",
+		Schema: `(
+			rig_id integer not null,
+			deal_id integer not null,
+			storage_provider text not null,
+			data_model_selector text not null,
+			deal_number integer not null,
+			primary key(rig_id, deal_id)
+		)`,
+	}
```
Comment (asutula) [inline pkg/storage/tableland/store.go:70]: Definition for the new deals Tableland table.

File: pkg/storage/tableland/store.go [hunk 29]
```diff
@@ -48,19 +56,24 @@ var (
...
-			layers_cid text,
-			image_full_name text,
-			image_full_alpha_name text,
-			image_medium_name text,
-			image_medium_alpha_name text,
-			image_thumb_name text,
-			image_thumb_alpha_name text,
-			animation_base_url text
+			label text not null,
+			value text not null
```
Comment (asutula) [inline pkg/storage/tableland/store.go:76]: Updated definition for the lookups Tableland table.

File: pkg/carstorage/nftstorage/client.go [hunk 30]
```diff
@@ -0,0 +1,223 @@
+package nftstorage
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply pkg/carstorage/nftstorage/client.go:1]: nice, odd they don't have a Go client!

**Review by sanderpick — APPROVED** [review]
Wow man, tons of work in here 💪 
LGTM, but I'll need some instructions on how to do the new publishing of tables, etc. 
We can put this in a ship a week or two from now.

**Comment by asutula** [comment]
Cool @sanderpick. I'll write a little playbook, and we can do the deployment together on Zoom. It's pretty simple actually. I guess I'll leave this unmerged until we decide on a ship week? Should we aim for next week?

**Comment by sanderpick** [comment]
sounds good. i'll update the linear project with a next week ship

**Pull request merged**